### PR TITLE
add -wo switch about use of obsolete features

### PR DESCRIPTION
**Pull request by WalterBright**
This is a proposed method for dealing with obsolete features. It's a switch `-wo` that warns about them. `-wo -we` will make them compilation errors.

This way, we can continue to compile legacy code without complaint, while enabling those who want to modernize their code an easy way to identify them. Since the switch is not the default, existing builds will still build.

It's not hooked up to anything yet, but could be hooked up to https://github.com/dlang/dmd/pull/15379 and https://github.com/dlang/dmd/pull/15380 and https://github.com/dlang/dmd/pull/15326

### CODE REVIEW

File: compiler/src/dmd/mars.d [hunk 1]
```diff
@@ -1942,6 +1942,8 @@ bool parseCommandLine(const ref Strings arguments, const size_t argc, ref Param
             params.warnings = DiagnosticReporting.error;
         else if (arg == "-wi")  // https://dlang.org/dmd.html#switch-wi
             params.warnings = DiagnosticReporting.inform;
+        else if (arg == "-wi")  // https://dlang.org/dmd.html#switch-wo
```
Comment (ibuclaw) [inline compiler/src/dmd/mars.d:1945]: Looks like a copy paste in error from previous condition.

**Review by ibuclaw — DISMISSED** [review]

File: compiler/src/dmd/mars.d (hunk 1 above)
Comment (WalterBright) [reply compiler/src/dmd/mars.d:1945]: yes it does, doesn't it!

**Comment by WalterBright** [comment]
`-w` does not make warnings into errors. Making `-wo` produce errors means the compiler will quit after a while, so one doesn't get them all. It should be best to print them all when someone just wants to scroll through the list and pick out the ones he wants to fix.

**Comment by ibuclaw** [comment]
> `-w` does not make warnings into errors. Making `-wo` produce errors means the compiler will quit after a while, so one doesn't get them all. It should be best to print them all when someone just wants to scroll through the list and pick out the ones he wants to fix.

Maybe allow users to pick which obsoleted feature they want to be warned about?

**Comment by WalterBright** [comment]
I expect there's going to be a lot of these, and few will bother with searching for specific ones.

**Comment by WalterBright** [comment]
What's up with buildkite dlang-tour/core ? The error doesn't seem related to this PR.

**Comment by WalterBright** [comment]
I need this to move forward on being legacy code compatible.

**Review by dkorpel — APPROVED** [review]

**Comment by ntrel** [comment]
> -w does not make warnings into errors

```
$ dmd --help |grep -- -w
  -w                warnings as errors (compilation will halt)
  -wi               warnings as messages (compilation will continue)
```

Also, there is no `-we` switch.

> Making -wo produce errors means the compiler will quit after a while, so one doesn't get them all

OK.

**Comment by WalterBright** [comment]
@ntrel yes you're right.

**Comment by Herringway** [comment]
Shouldn't this have had a changelog entry?

**Pull request merged**